### ISSUE TITLE
blob/azureblob: support using AAD credentials for default URLOpener

### ISSUE THREAD

**Issue by omerlh** [google/go-cloud#2410]
Please use a title starting with the name of the affected package, or \"all\",
followed by a colon, followed by a short summary of the feature request.
Example: `blob/gcsblob: add support for more blobbing`.

### Is your feature request related to a problem? Please describe.
Azure storage now supports authentication using [Azure Active Directory](https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad) user, which is easier and more secure compared to storage account key or SAS tokens.

### Describe the solution you'd like

Be able to use active directory credentials instead of SAS or access key to login to blob.

**Comment by vangent** [comment]
I believe this should be possible already, using [`azblob.NewTokenCredential`](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#NewTokenCredential). The `azureblob` documentation is hard to follow, but something like this:

```
import "github.com/Azure/azure-storage-blob-go/azblob"
import "gocloud.dev/blob/azureblob"

...

// First, get an oauth2 token from AAD.
// There are examples here:
// https://docs.microsoft.com/en-us/go/azure/azure-sdk-go-auth-aad-samples.

// Use it to create an azblob.Credentials, which can be used to make a
// pipeline for opening a bucket.
cred := azblob.NewTokenCredential(token, nil /* or add a refresher */)
pipeline := azureblob.NewPipeline(cred, azblob.PipelineOptions{})
bucket, err := azureblob.OpenBucket(ctx, pipeline, accountName, containerName, new(azureblob.Options)) 

// ... use your bucket!
```

I don't think it's feasible to support this kind of authentication using an `URLOpener`. Please advise if you think otherwise.

**Comment by omerlh** [comment]
Well, it's relying on environment variables today for authentication (which is kinda not the best practices anyway). It's possible to add AAD authentication based on env var - maybe even with Azure managed identity.

**Comment by vangent** [comment]
OK. I'm going to retitle this issue to include "in URL openers" since it is supported using constructors, and put it to Unplanned.

**Comment by vangent** [comment]
Also: note that you can create a custom [`URLOpener`](https://godoc.org/gocloud.dev/blob/azureblob#URLOpener) and set the `Pipeline` field after you've authenticated, and you can use that `URLOpener` to open URLs. The environment variable approach is used in the default URLOpener.

**Comment by dadgar** [comment]
> I believe this should be possible already, using [`azblob.NewTokenCredential`](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#NewTokenCredential). The `azureblob` documentation is hard to follow, but something like this:
> 
> ```
> import "github.com/Azure/azure-storage-blob-go/azblob"
> import "gocloud.dev/blob/azureblob"
> 
> ...
> 
> // First, get an oauth2 token from AAD.
> // There are examples here:
> // https://docs.microsoft.com/en-us/go/azure/azure-sdk-go-auth-aad-samples.
> 
> // Use it to create an azblob.Credentials, which can be used to make a
> // pipeline for opening a bucket.
> cred := azblob.NewTokenCredential(token, nil /* or add a refresher */)
> pipeline := azureblob.NewPipeline(cred, azblob.PipelineOptions{})
> bucket, err := azureblob.OpenBucket(ctx, pipeline, accountName, containerName, new(azureblob.Options)) 
> 
> // ... use your bucket!
> ```
> 
> I don't think it's feasible to support this kind of authentication using an `URLOpener`. Please advise if you think otherwise.

@vangent I think it would be great to add this to the docs as it really saved me a lot of time. If you want to link a concrete implementation, I used your links to come up with the following:

```
	// Get an Oauth2 token for the account for use with Azure Storage
	ccc := auth.NewClientCredentialsConfig(azCreds.ClientID, azCreds.ClientSecret, azAccount.TenantId)
	ccc.Resource = azure.PublicCloud.ResourceIdentifiers.Storage
	token, err := ccc.ServicePrincipalToken()
	if err != nil {
		return nil, err
	}

	// Failed to refresh OAuth2 token
	if err := token.RefreshWithContext(ctx); err != nil {
		return nil, err
	}

	blobcred := azblob.NewTokenCredential(token.OAuthToken(), nil)
	pipeline := azureblob.NewPipeline(blobcred, azblob.PipelineOptions{})
	return azureblob.OpenBucket(ctx, pipeline, azureblob.AccountName(accountName), containerName, new(azureblob.Options))
```

**Comment by vangent** [comment]
That sounds reasonable as an example. Would you mind sending a PR to add an example to https://github.com/google/go-cloud/blob/master/blob/azureblob/example_test.go ?

**Comment by vangent** [comment]
Fixed!
